tensorflow_model_optimization/python/core/sparsity_tf2/riglpruner_test.py

```diff
@@ -0,0 +1,577 @@
+    del result1, result2
+    del expected1, expected2, mask, optimizer
+
+
+    weight1 = tf.Variable(np.linspace(1.0, 100.0, 10), dtype=tf.float32)
+    mask = tf.constant([1., 1., 0., 1., 0., 0., 1., 1., 0., 0.], dtype=weight_dtype)
+    grad = self.grad(weight1)
+    sparse_vars = [(mask, weight1, grad)]
+    _optimizer = tf.keras.optimizers.SGD(learning_rate=0.01)
+    optimizer = pruning_optimizer.PruningOptimizer(_optimizer, sparsity_config)
+    optimizer.iterations.assign(0)
+    p.create_slots(optimizer, weight1)
+    # assign to the correct initialization mask
+    optimizer.get_slot(weight1, 'mask').assign(mask)
+
+    expected1 = tf.constant([0., 0., 1., 1., 1., 0., 1., 1., 0., 0.])
+    expected2 = tf.constant([0., 1., 1., 0., 1., 0., 1., 1., 0., 0.])
+    result1, result2 = tf.function(_train)(optimizer, weight1)
+    self.assertAllEqual(result1['sparsity'], self.target_sparsity)
+    self.assertAllEqual(result2['sparsity'], self.target_sparsity)
+    # # assert that sparsity does not change
+    self.assertAllEqual(tf.reduce_sum(result1['before']), tf.reduce_sum(result1['after']))
+    self.assertAllEqual(tf.reduce_sum(result2['before']), tf.reduce_sum(result2['after']))
+    # # assert there are some changes in the mask during each update
+    self.assertAllEqual(expected1, result1['after'])
+    self.assertAllEqual(result1['after'], result2['before'])
+    self.assertAllClose(expected2, result2['after'])
+
+
+  @parameterized.parameters(
+    (0.5,), (0.7,), (0.5,), (0.3,),
+  )
+  def testUpdatesAccordingtoSchedule(self, drop_ratio):
+    # unlike previous tests, this checks that the masked weights
+    # maintain consistent sparsity and that there is some change in the mask.
+    # weight = tf.Variable(np.linspace(1.0, 100.0, 100))
+    weight = tf.Variable(np.linspace(1.0, 100.0, 100), dtype=tf.float32)
+    weight_dtype = weight.dtype.base_dtype
+    mask = tf.Variable(
+        tf.ones(weight.get_shape(), dtype=weight_dtype),
+        dtype=weight_dtype)
+    grad = self.grad(weight)
+    sparse_vars = [(mask, weight, grad)]
+    sparsity_params = {
+      'pruning_schedule': self.updater(drop_ratio, 1, 4, 2)
+    } # dummy just for using the pruning optimizer
+    sparsity_config = pruning_config.LowMagnitudePruningConfig(**sparsity_params) # dummy
+
+    p = pruner.RiGLPruner(
+      update_schedule=self.updater(drop_ratio, 1, 4, 2), # update iter 1 and 3
+      sparsity=self.target_sparsity,
+      block_size=self.block_size,
+      block_pooling_type=self.block_pooling_type,
+      seed=self.seed,
+      noise_std=self.noise_std,
+      reinit=self.reinit
+    )
+
+    _optimizer = tf.keras.optimizers.SGD(learning_rate=0.01)
+    optimizer = pruning_optimizer.PruningOptimizer(_optimizer, sparsity_config)
+    optimizer.iterations.assign(0)
+
+    p.create_slots(optimizer, weight)
+
+    def _train(optimizer, weight):
+      expected_iter1 = {'before': tf.zeros_like(weight, dtype=weight.dtype.base_dtype) * -1,
```

Comment:
This test is a bit hard-coded as written. Maybe write the values of the mask & the sparsity to a list (where index = the current step) and then return the list at the end? The checks would be much easier to write when not needing to consider before vs. after vs. hard-coding the iterations in this `_train` method.

To get around the autograph limitation it's okay to switch the `tf.range(0, 5)` below to just a `range(0, 5)`